airbyte-cdk/bulk/core/load/src/main/kotlin/io/airbyte/cdk/message/DestinationMessageDeserializer.kt

```diff
@@ -16,6 +16,9 @@ interface Deserializer<T : Any> {
  * Converts the internal @[AirbyteMessage] to the internal @[DestinationMessage] Ideally, this would
  * not use protocol messages at all, but rather a specialized deserializer for routing.
  */
+// Does this _need_ to be injected?
+// Non-rethorical question: are we going to have different deserialization schemes?
+// This comment applies to a number of similar definitions and so I won't repeat it.
```

Comment:
There are two reasons I've tried to err on the of injecting most dependencies.

One is for the convenience for unit testing. The more everything is swappable/parameterized, the easier it is to sub in simple mocks.

The other is because in general the reason why rewrites fail is because of faulty assumptions about what ought to be coupled/decoupled. More interfaces means less coupling. More interfaces w/o DI means more plumbing, which actually means coupling between object instantiations (and creating an implicit hierarchy of visibility and control that is itself based on assumptions that might change). The idea that it will be trivial to refactor and decouple later as needed is a fallacy disproven by the fact that we're rewriting.

Also in this specific case, yes, we might have more than one deserializer. Some destinations don't need the client data ever to be deserialized, so a possible optimization for them is a custom deserializer that only parses the metadata. (A lot of pain in the existing CDK is from juggling optimizations made to the deserialization path for some destinations that don't work for others.)

Reply:
FWIW I buy your second argument, but perhaps not your first. While ease of writing tests _in connectors_ absolutely is something worth optimizing for, ease of writing tests is _not_ worth optimizing for in the CDK if the CDK is going to be a slowly-changing body of code, which will be the case if we're successful in this endeavour. In that case, readability and predictability, which implies a lack of magic, takes precedence. 

This is not at odds with DI per se but let's be real, if you're injecting 10 things you can reason about all of them and their interdependencies, but if you're injecting 50+ you just can't.

Your argument about coupling is pretty convincing. I guess the conclusion I'm coming to is that it might be worth doing a pass once the CDK stabilizes to refactor by auditing each bean and asking ourselves if it really needs to exist in the first place.

Reply:
I agree. After reading your first round of comments, I added "do a micronaut simplification pass" to my release todo list. A lot of this was me learning what is/isn't necessary. Like the DestinationTaskLauncher <-> TaskFactory decoupling is almost definitely overkill, because the specific task launcher implementations are workflows, and workflow <-> task is never going to change generically.

Reply:
I think I disagree though that it's possible for this to become a slow-changing codebase. At least on the destination side, most common features will end up living in the CDK, and features never stop. Also we may eventually have divergent workflow patterns (like cloud storage v dbs).